File: problems/字符串/186.翻转字符串里的单词.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void reverseWords(vector<char>& s) {
    reverse(s.begin(), s.end());
    int i = 0;
    for (int j = 0; j <= s.size(); j++) {
        if (j == s.size() || s[j] == ' ') {
            reverse(s.begin() + i, s.begin() + j);
            i = j + 1;
        }
    }
}

int main() {
    // 测试用例
    vector<char> t1 = {'t', 'h', 'e', ' ', 's', 'k', 'y', ' ', 'i', 's', ' ', 'b', 'l', 'u', 'e'};
    reverseWords(t1);
    for (char c : t1) cout << c;
    cout << endl;  // 预期: blue is sky the

    vector<char> t2 = {' ', ' ', 'h', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd', ' ', ' '};
    reverseWords(t2);
    for (char c : t2) cout << c;
    cout << endl;  // 预期: world hello
    return 0;
}

File: problems/字符串/186.翻转字符串里的单词.py
#!/usr/bin/env python3
"""
186. 翻转字符串里的单词 II
"""


def reverse_words(s: list[str]) -> None:
    s.reverse()
    i = 0
    for j in range(len(s) + 1):
        if j == len(s) or s[j] == ' ':
            l, r = i, j - 1
            while l < r:
                s[l], s[r] = s[r], s[l]
                l += 1
                r -= 1
            i = j + 1


if __name__ == "__main__":
    # 测试用例
    s1 = list("the sky is blue")
    reverse_words(s1)
    print(''.join(s1))  # 预期: "blue is sky the"

    s2 = list("  hello world  ")
    reverse_words(s2)
    print(''.join(s2))  # 预期: "world hello"
